beacon/sync/src/main/java/tech/pegasys/teku/beacon/sync/historical/ReconstructHistoricalStatesService.java

```diff
@@ -93,12 +102,33 @@ protected SafeFuture<?> doStart() {
 
   public void applyBlocks(final BeaconState genesisState, final UInt64 anchorSlot) {
     Context context = new Context(genesisState, SpecConfig.GENESIS_SLOT.plus(1), anchorSlot);
-    applyNextBlock(context).finish(statusLogger::reconstructHistoricalStatesServiceFailedProcess);
+    applyNextBlock(context)
+        .finish(
+            error -> {
+              final Throwable rootCause = Throwables.getRootCause(error);
+              if (rootCause instanceof ShuttingDownException
+                  || rootCause instanceof InterruptedException
+                  || rootCause instanceof RejectedExecutionException) {
+                LOG.debug("Shutting down", rootCause);
+              } else {
+                statusLogger.reconstructHistoricalStatesServiceFailedProcess(error);
+              }
+            });
   }
 
   private SafeFuture<Void> applyNextBlock(Context context) {
     if (context.checkStopApplyBlock()) {
       statusLogger.reconstructHistoricalStatesServiceComplete();
+      stopped.complete(null); // fixme should this be set here too?
```

Comment:
Yep, we want to call complete here. That way if the service finishes it's work before stop() is called, the future returned by stop() will be complete - otherwise it would never complete because `applyNextBlock` wouldn't be called anymore and so we'd never check the shutDown boolean again.